modules/internal/LocaleModelHelpAPU.chpl

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2017 Advanced Micro Devices, Inc.
+ * Copyright 2004-2017 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+module LocaleModelHelpAPU {
+
+  param localeModelHasSublocales = true;
+
+  use LocaleModelHelpSetup;
+  use LocaleModelHelpRuntime;
+
+  //////////////////////////////////////////
+  //
+  // utilities
+  //
+  inline
+  proc chpl_getSubloc() {
+    extern proc chpl_task_getSubloc(): chpl_sublocID_t;
+    return chpl_task_getSubloc();
+  }
+
+  //////////////////////////////////////////
+  //
+  // support for "on" statements
+  //
+
+  //
+  // runtime interface
+  //
+  extern proc chpl_task_setSubloc(subloc: int(32));
+
+  //
+  // returns true if an executeOn can be handled directly
+  // by running the function in question.
+  // Applies to execute on and execute on fast.
+  // When performing a blocking on, the compiler will emit this sequence:
+  //
+  //  if (chpl_doDirectExecuteOn(targetLocale))
+  //         onStatementBodyFunction( args ... );
+  //  else
+  //         chpl_executeOn / chpl_executeOnFast
+  //
+  export
+  proc chpl_doDirectExecuteOn(loc: chpl_localeID_t // target locale
+                             ):bool {
+    const dnode =  chpl_nodeFromLocaleID(loc);
+    const dsubloc =  chpl_sublocFromLocaleID(loc);
+
+    if (dnode != chpl_nodeID) {
+      return false; // need to move to different node
+    } else {
+      var origSubloc = chpl_task_getRequestedSubloc();
+      if (dsubloc==c_sublocid_any || dsubloc==origSubloc) {
+        return true;
+      } else {
+        return false; // need to move to different sublocale
+      }
+    }
+  }
+
+  //
+  // regular "on"
+  //
+  pragma "insert line file info"
+  export
+  proc chpl_executeOn(loc: chpl_localeID_t, // target locale
+                      fn: int,              // on-body function idx
+                      args: chpl_comm_on_bundle_p,     // function args
+                      args_size: size_t     // args size
+                     ) {
+    const dnode =  chpl_nodeFromLocaleID(loc);
+    const dsubloc =  chpl_sublocFromLocaleID(loc);
+    if (dsubloc == 0) {
+        warning("executing on CPU");
```

Comment:
I'd expect such printouts to be controlled by a config param (maybe they are and I'm missing something?)

Reply:
Also, I don't see so much different here from LocaleModelHelpNuma (but I might be missing something). Makes me wonder if the APU model could use LocaleModelHelpNuma. That would cause it to use more regularly tested code. I'm not too worried about it either way, though - philosophically the APU locale model needs to specify these functions and could do so by implementing them from scratch. I'd just like to avoid copy/paste code we have to maintain if we can.

Reply:
I actually meant to remove those printouts. I'll consider putting them under some config param though. Your suggestion of making it just use LocaleModelHelpNuma is an interesting one (for this and the function in LocaleModelHelpSetup). I'll have to think about this a little bit more.